apps/opencs/view/render/terraintexturemode.cpp

```diff
@@ -372,21 +444,99 @@ void CSVRender::TerrainTextureMode::editTerrainTextureGrid(const WorldspaceHitRe
+
+        if(allowLandTextureEditing(mCellId)==true && !mCustomBrushShape.empty())
+        {
+            for(auto const& value: mCustomBrushShape)
+            {
+                if(yHitInCell + value.second >= 0 && yHitInCell + value.second <= 15 && xHitInCell + value.first >= 0 && xHitInCell + value.first <= 15)
+                {
+                    newTerrain[(yHitInCell+value.second)*landTextureSize+xHitInCell+value.first] = brushInt;
+                }
+                else
+                {
+                    int cellXDifference = std::floor(1.0f*(xHitInCell + value.first)/landTextureSize);
+                    int cellYDifference = std::floor(1.0f*(yHitInCell + value.second)/landTextureSize);
+                    int xInOtherCell = xHitInCell + value.first - cellXDifference * landTextureSize;
+                    int yInOtherCell = yHitInCell + value.second - cellYDifference * landTextureSize;
+
+                    std::string cellId = CSMWorld::CellCoordinates::generateId(cellX+cellXDifference, cellY+cellYDifference);
+                    if (allowLandTextureEditing(cellId)==true)
+                    {
+                        CSMWorld::LandTexturesColumn::DataType newTerrainPointerOtherCell = landTable.data(landTable.getModelIndex(cellId, textureColumn)).value<CSMWorld::LandTexturesColumn::DataType>();
+                        CSMWorld::LandTexturesColumn::DataType newTerrainOtherCell(newTerrainPointerOtherCell);
+                        newTerrainOtherCell[yInOtherCell*landTextureSize+xInOtherCell] = brushInt;
+                        pushEditToCommand(newTerrainOtherCell, document, landTable, cellId);
+                    }
                 }
             }
+            pushEditToCommand(newTerrain, document, landTable, mCellId);
+        }
+    }
+}
+
+void CSVRender::TerrainTextureMode::selectTerrainTextures(std::pair<int, int> texCoords, unsigned char selectMode, bool dragOperation)
+{
+    int r = mBrushSize / 2;
+    std::vector<std::pair<int, int>> selections;
+
+    if (mBrushShape == 0)
+    {
+        selections.emplace_back(texCoords);
+    }
+
+    if (mBrushShape == 1)
+    {
+        for(int i = texCoords.first - r; i <= texCoords.first + r; ++i)
+        {
+            for(int j = texCoords.second - r; j <= texCoords.second + r; ++j)
+            {
+                selections.emplace_back(std::make_pair(i, j));
+            }
+        }
+    }
+
+    if (mBrushShape == 2)
+    {
+        for(int i = texCoords.first - r; i <= texCoords.first + r; ++i)
+        {
+            for(int j = texCoords.second - r; j <= texCoords.second + r; ++j)
+            {
+                int distanceX = abs(i - texCoords.first);
+                int distanceY = abs(j - texCoords.second);
+                int distance = std::round(sqrt(pow(distanceX, 2)+pow(distanceY, 2)));
+                if (distance < r) selections.emplace_back(std::make_pair(i, j));
```

Comment:
I noticed of course that this syntax is used elsewhere, but I wonder if these loops could be rewritten like
```cpp
        for (int i = -r; i <= r; i++)
        {
            for (int j = -r; j <= r; j++)
            {
                osg::Vec2f coords(i,j);
                if (std::round(coords.length()) < r)
                    selections.emplace_back(i + texCoords.first, j + texCoords.second);
            }
        }
```

Reply:
Probably they could. That looks okay. I wrote it on multiple lines in this syntax to cut the process into parts, to make formulas easier to understand, but I don't think your proposition is unclear. Umm.. Is this the way how pairs can be made with emplace_back?

Reply:
Yes.